common4j/src/main/com/microsoft/identity/common/java/util/CommonURIBuilder.java

```diff
@@ -0,0 +1,162 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.util;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+
+import cz.msebera.android.httpclient.NameValuePair;
+import cz.msebera.android.httpclient.client.utils.URIBuilder;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+/**
+ * Our URIBuilder.
+ * We want to make sure we never send duplicated parameters to the server.
+ * This is done by
+ * 1. disabling {@link cz.msebera.android.httpclient.client.utils.URIBuilder#addParameter(String, String)} and
+ *    {@link cz.msebera.android.httpclient.client.utils.URIBuilder#addParameters(List))}
+ * 2. adding {@link CommonURIBuilder#addParametersIfAbsent}
+ */
+public class CommonURIBuilder extends cz.msebera.android.httpclient.client.utils.URIBuilder {
+
+    public CommonURIBuilder(){
+        super();
+    }
+
+    public CommonURIBuilder(final String string) throws URISyntaxException {
+        super(string);
+    }
+
+    public CommonURIBuilder(final URI uri) {
+        super(uri);
+    }
+
+    @Override
+    public CommonURIBuilder addParameters(@NonNull final List<NameValuePair> nvps) {
+        throw new UnsupportedOperationException("This should never be used. Either use setParameter or addParametersIfAbsent");
+    }
+
+    @Override
+    public CommonURIBuilder addParameter(@NonNull final String param, @NonNull final String value) {
+        throw new UnsupportedOperationException("This should never be used. Either use setParameter or addParametersIfAbsent");
+    }
+
+    @Override
+    public CommonURIBuilder setParameters(List<NameValuePair> nvps) {
+        super.setParameters(nvps);
+        return this;
+    }
+
+    @Override
+    public CommonURIBuilder setParameters(NameValuePair... nvps) {
+        super.setParameters(nvps);
+        return this;
+    }
+
+    @Override
+    public CommonURIBuilder setParameter(String param, String value) {
+        super.setParameter(param, value);
+        return this;
+    }
+
+    /**
+     * Adds parameters to URI query if it does not already exist.
+     * The parameter name and value are expected to be unescaped and may contain non ASCII characters.
+     * <p>
+     * Please note query parameters and custom query component are mutually exclusive. This method
+     * will remove custom query if present.
+     * </p>
+     *
+     * @param params list of parameters.
+     * @return {@link CommonURIBuilder}
+     */
+    public CommonURIBuilder addParametersIfAbsent(@Nullable final Map<String, ?> params) {
```

Comment:
Can we make this follow the pattern from the superclass?

Reply:
That is to say

```
   public URIBuilder addParametersIfAbsent(final List <NameValuePair> nvps) {
 ```
and

```
   public URIBuilder addParametersIfAbsent(final NameValuePair... nvps) {
 ```

Reply:
will do. This is an extension from that though.

Reply:
I know... just asking to use the list of NameValuePair instead of Map.

Reply:
@AdamBJohnsonx I'm inclined not to add these two method... because
1. They're not being used. adding methods means more maintainance (adding tests, etc). If we need to use them (which I doubt), then we can always add them later.
2. List <NameValuePair> has the same type erasure as List<Map.Entry<String, String>>